File: client/src/main/java/collection/CollectionFilter.java
package collection;

import model.data.Flat;
import model.data.Model;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class CollectionFilter {
  private Set<Flat> objectsCollection;
  private Long collectionVersion;
  private LocalDateTime creationDateTime;
  private final ChangesApplier<Flat> changesApplier;

  public CollectionFilter(CollectionWrapper<Flat> collectionWrapper) {
    loadFullCollection(collectionWrapper);
    changesApplier = new ChangesApplier<>(objectsCollection);
  }

  public void loadFullCollection(CollectionWrapper<Flat> collectionWrapper) {
    objectsCollection = new HashSet<>(collectionWrapper.getCollection());
    creationDateTime = collectionWrapper.getCreationDateTime();
    collectionVersion = collectionWrapper.getVersion();
  }

  public void applyChangelist(CollectionChangelist<Flat> changelist) {
    collectionVersion = changelist.getLatestVersion();
    changesApplier.apply(changelist);
  }

  /**
   * Returns collection creation time
   */
  public LocalDateTime getCreationDateTime() {
    return creationDateTime;
  }

  /**
   * Returns collection version. Used in logic clock syncronization.
   */
  public long getCollectionVersion() {
    return collectionVersion;
  }

  /**
   * Returns collection size.
   *
   * @see LinkedHashSet#size()
   */
  public int getSize() {
    return objectsCollection.size();
  }

  /**
   * Returns array of {@link Flat} which name contains passed string.
   *
   * @param filter string to lookup in names.
   */
  public Model[] filterContainsName(String filter) {
    return objectsCollection.stream().sorted().filter(flat -> flat.getName().contains(filter)).toArray(Model[]::new);
  }

  /**
   * Returns a {@link Set} of unique {@link Long} values got by {@link Flat#getNumberOfRooms()} from each entry.
   */
  public Set<Long> getUniqueNumberOfRooms() {
    return objectsCollection.stream().map(Flat::getNumberOfRooms).collect(Collectors.toSet());
  }

  /**
   * Returns an array of {@link Boolean} values in descending order got by {@link Flat#getNewness()} from each entry.
   */
  public Boolean[] getFieldDescendingNew() {
    return objectsCollection.stream().sorted(Comparator.reverseOrder()).map(Flat::getNewness).toArray(Boolean[]::new);
  }
}

File: client-ui/domain/src/commonMain/kotlin/ru.moleus.kollector/domain/collection/CollectionFilter.kt
package ru.moleus.kollector.domain.collection

import collection.CollectionChangelist
import collection.CollectionWrapper
import com.arkivanov.decompose.value.MutableValue
import com.arkivanov.decompose.value.ValueObserver
import com.arkivanov.decompose.value.reduce
import model.data.Flat
import model.data.Model
import java.time.LocalDateTime

class CollectionFilter(collectionWrapper: CollectionWrapper<Flat>) {
    private val objectsCollection: MutableSet<Flat> = HashSet(collectionWrapper.collection)
    private val collectionValue = MutableValue(objectsCollection)

    val collection: Set<Flat> = objectsCollection
    private var collectionVersion = collectionWrapper.version

    /**
     * Returns collection creation time
     */
    val creationDateTime: LocalDateTime = collectionWrapper.creationDateTime
    private val changesApplier: ChangesApplier<Flat> = ChangesApplier(target = objectsCollection)

    fun subscribeOnUpdate(onUpdate: ValueObserver<Set<Flat>>) {
        collectionValue.subscribe { collection -> onUpdate(collection) }
    }

    fun applyChangelist(changelist: CollectionChangelist<Flat>) {
        collectionVersion = changelist.latestVersion
        changesApplier.apply(changelist = changelist)
        collectionValue.reduce { objectsCollection }
    }

    /**
     * Returns collection version. Used in logic clock syncronization.
     */
    fun getCollectionVersion(): Long {
        return collectionVersion
    }

    /**
     * Returns collection size.
     *
     * @see LinkedHashSet.size
     */
    val size: Int
        get() = objectsCollection.size

    /**
     * Returns array of [Flat] which name contains passed string.
     *
     * @param filter string to lookup in names.
     */
    fun filterContainsName(filter: String): Array<Model> =
        objectsCollection.filter { flat -> flat.name.contains(filter) }.toTypedArray()

    /**
     * Returns a [Set] of unique [Long] values got by [Flat.getNumberOfRooms] from each entry.
     */
    val uniqueNumberOfRooms: Set<Long>
        get() = objectsCollection.map<Flat, Long> { it.numberOfRooms }.toSet()

    /**
     * Returns an array of [Boolean] values in descending order got by [Flat.getNewness] from each entry.
     */
    val fieldDescendingNew: Array<Boolean>
        get() = objectsCollection.reversed().map<Flat, Boolean> { it.newness }.toTypedArray()
}